### ISSUE TITLE
SPIKE: Convert BRJS to use SystemJS

### ISSUE THREAD

**Issue by dchambers** [BladeRunnerJS/brjs#1389]
I've been working on three separate design proposals, all of which are affected by the need for a common bundling format that can also support something akin to sub-realms:

  * [NPM Design Proposal](https://github.com/BladeRunnerJS/brjs/wiki/BladeRunnerJS-NPM-Integration-Strawman) (requires unified bundling format)
  * [ES6 Design Proposal](https://github.com/BladeRunnerJS/brjs/issues/1303) (requires unified bundling format)
  * [New Service Registry Design Proposal](https://github.com/BladeRunnerJS/brjs/wiki/New-Registry-Proposal) (dependent on whether our new module system can allow modules to be mocked in test)

I've so far been able to prove within a spike that [SystemJS](https://github.com/systemjs/systemjs) fulfills all of our needs, and appears to be supported by the transpilers we might want to use to add support for ES6. The next thing to do is to convert BRJS to use SystemJS within a branch to further validate these three design proposals before any work is done.

This requires the following tasks to be performed:

  - [x] Verify that I can use babelify to generate named modules that can be consumed by SystemJS (within the [es6-module-loader-spike](https://github.com/dchambers/es6-module-loader-spike)).
  - [x] Convert BRJS to use SystemJS (verify the app runs).
  - [ ] Implement a [MockableSystem](https://github.com/whatwg/loader/issues/30#issuecomment-101567409) library and update sub-realms to proxy through to it (verify the tests run).

**Comment by dchambers** [comment]
Problems with babelify's `System.register()` output so far:

  1. It doesn't emit named modules that can be bundled by default (it uses anonymous modules that can be loaded one at a time),  but this can be fixed by using the `--moduleIds true` option.
  1. It uses absolute paths for named modules, whereas we'd want relative names (this could be hacked from the outside).
  1. It seems to only generate all source code if you use a path that causes it to find all JavaScript, but then it uses a System.register() for each module rather than the main module for each library, which means there could potentially be name clashes &mdash; particularly since the main module may not be located in the top-level directory.
  1. Having to use a wildcard to find all source would also make it bundle modules that were never actually used.

If I use the CommonJS module formatter instead (`--modules common` instead of `--modules system`) then it seems to generate output that is much closer to what I would expect, accept that it is no longer in `System.register()` format. Given that Babel supports [custom module formatters](https://babeljs.io/docs/usage/modules/#custom), this looks like the next thing to investigate...

**Comment by dchambers** [comment]
Yesterday I discovered that [jspm](http://jspm.io/) includes a bundler that works exactly as we need, plus has some other nice features that might come in handy. I haven't yet considered the implications of using jspm for the various design proposals yet though.

**Comment by dchambers** [comment]
Yesterday I was able to [spike](https://github.com/dchambers/es6-module-loader-spike) sub-realm like support using jspm and a [couple](https://github.com/dchambers/es6-module-loader-spike/blob/master/index.html#L22) [of](https://github.com/dchambers/es6-module-loader-spike/blob/master/index.html#L41) bug-fixes to SystemJS that have yet to be released. Plus I'd previously spiked a sub-realm like feature using a more standards based approach, that wouldn't be tied only to SystemJS.

Here are my findings:

**Standards Based Approach:**
  1. We use an instantiate hook to keep a record of each module's dependencies.
  1. We allow one or more modules to be temporarilly replaced (by invoking `delete()` and `set()`).
  1. We re-import all dependent modules (including transitive dependencies) from the server using `import()`.

**SystemJs Based Approach:**
  1. We replace `System.register` with an implementation that maintains a map of all of the modules.
  1. We include a mechanism for allowing modules to be replaced, causing all dependent modules to be re-registered (by invoking `System.register()` again).

**Conclusion:**

The standards based approach would normally be the best option, but in this case there are a number of reasons it's not as good:

  1. Tests would run much slower if a torrent of network activity had to occur every time a test mocked a module, even when running over `localhost`.
  1. We would need a server that could serve up individual modules so they could be replaced using `System.import()`, and there's no convenient way of doing this in js-test-driver.
  1. The ES6 Module Loader has been removed from ES2015, and the version that will be included into ES2016 already looks quite different to the polyfills that still have the original ES6 design, so any work done now won't work natively on browsers in the future anyway.

**Comment by dchambers** [comment]
I've just created the `README.md` for a potential new [MockableSystemJS](https://github.com/dchambers/MockableSystemJS) library to replace [browser-modules](https://github.com/BladeRunnerJS/browser-modules). @james-shaw-turner / @andyberry88, can you take a look please?

**Comment by dchambers** [comment]
I've been thinking about the fact that jspm builds exactly the type of bundles we'd need if we were willing to have a hard dependency on Node.js, and that it can also work with pre-built bundles for users that aren't willing to have a hard dependency on Node.js. However, it's been pointed out that the use of pre-built bundles would cause apps to load lots of libraries and modules they don't actually use.

On the other hand, if we're going to process either the `node_modules` or the `jspm_packages` directory ourselves, than `jspm_packages` is a bad idea because:

  1. It's harder to interpret than the `node_modules` directory format since it depends on `map` and `paths` values which are set within JavaScript rather than config (typically within `config.js`).
  1. It doesn't allow libraries to be distributed in different locations the way they can be with `node_modules`.
  1. Although it can allow local libraries to be developed outside `jspm_packages`, this is again specified in code.

Therefore, if we don't want to depend on Node.js, we'd need to be able to parse `package.json` files to determine the libraries each library depends on (this step could potentially even be omitted), and be able to scan upwards looking for `node_modules` directories that contain that module. If we found a module, then we would need to do the same again for it's dependencies, etc.

Additionally, we'd need to wrap each module within each library into a `System.register()` invocation, but where a local `require()` function is provided to access the map of imports provided to the module, and which otherwise falls back to `System.get()` for modules that weren't determined through static analysis.

**Comment by dchambers** [comment]
After speaking to @james-shaw-turner earlier today I realized that it's worth documenting the various NPM bundling options:

  1. [browserify](http://browserify.org/): this produces a CommonJS style bundle which involves defining `require()`, making it a little hard to combine with [browser-modules](https://github.com/BladeRunnerJS/browser-modules), which does the same, and also meaning that the sub-realms feature would only allow BRJS modules to be mocked, but not NPM modules.
  1. [babelify](https://github.com/babel/babelify): although this can be used to transpile ES6, it is also of interest here because it supports bundling to [`System.register`](https://github.com/ModuleLoader/es6-module-loader/wiki/System.register-Explained) format, which allows bundles from different places to be more easily combined. Unfortunately, it outputs anonymous bundles which are designed for on-demand loading on a server rather than pre-emptive bundling.
  1. [jspm](http://jspm.io/): This also produces `System.register` bundles, and produces bundles designed for sending to the browser, but requires `jspm install` to be used instead of `npm install`.

**Comment by dchambers** [comment]
Also, after speaking to @james-shaw-turner we decided the following questions needed to be answered before we could contemplate processing `node_modules` ourselves:

  1. Would we really need to process 'package.json' to discover the external libraries each library depends on?
  1. How does jspm handle inner modules &mdash; does this require something like browserify instead of using `System.register` for these?

**Comment by janhancic** [comment]
There's also [webpack](http://webpack.github.io/) (just to make your list complete).

**Comment by dchambers** [comment]
@janhancic, true.

I guess I never bothered investigating webpack since I was looking for something to bundle NPM modules, and browserify and babelify where the natural first choices for that. Since then, I ended up honing in on JSPM / SystemJS / ES6 Module Loader since it allows different typed modules to inter-operate, and since it's compatible with our existing _sub-realms_ feature.

Additionally, the ES6 Module Loader API allows CSS, HTML, etc to be bundled in much the same way that webpack does, but using a neater _transformation-hook_, rather than sending these assets as JavaScript over the wire.

**Comment by dchambers** [comment]
Today I created a [jspm-spike](https://github.com/dchambers/jspm-spike) repository that provides a very simple set of inter-dependent NPM modules so we can see [the resultant bundle](https://github.com/dchambers/jspm-spike/blob/master/JSPMLib/build.js) that JSPM itself would produce.

Fortunately, it turns out the output is actually very easy to understand, and is certainly much simpler than the output browserify produces. Best of all, it solves the problem of needing a hierarchical module system to segregate internal and external dependencies, and does this in a really neat way:

  1. All modules, both internal and external, are made available within a single module loader.
  1. Name clashes are solved by prefixing the external module name to each module, where a map (configured using `System.config()`) is used to map the external module names to their _fully-qualified_ form.
  1. Soft encapsulation is used, in that it now becomes possible to request or mock a modules inner modules if one is willing to use the _fully-qualified_ names, but this is actually nice since it needs to be a deliberate action, but provides useful flexibility NPM doesn't naturally offer.

I've still yet to consider whether we actually need to process `package.json` yet, but this is a minor point, and it now looks very feasible for us to continue having a soft dependency on Node.js if that's what we want.

**Comment by andyberry88** [comment]
> I've just created the README.md for a potential new MockableSystemJS library to replace browser-modules. @james-shaw-turner / @andyberry88, can you take a look please?

Looks good @dchambers :+1:

**Comment by dchambers** [comment]
I updated the jspm-spike to include an example of a version fork (look for `asap` in [config.js](https://github.com/dchambers/jspm-spike/blob/master/JSPMLib/config.js)), and this shows why we would need to process `package.json` after all, since otherwise we'd end up with an unnecessarily bloated `config.js`.

**Comment by dchambers** [comment]
Notes for bringing the team up to speed with the results of my NPM investigation:

  1. Explain how Node.js works (`require()` + `node_modules`).
  2. Explain how NPM works (`package.json` -> `node_modules`).
  3. Explain how NPM allows the same set of libraries used to create programs to also create tooling (`devDependencies`, `npm install` & `prepublish`)
  4. Explain how Node.js modules are expected to be CommonJS / ES5.
  5. Explain how people create ES6 NPM modules.
  6. Explain how NPM modules are primarilly designed for use on the server, but how people use `browserify` and `babelify` to allow those same libraries to be used on the browser, including by non-NPM users by pushing browserified bundles to `dist`).
  7. Explain how JSPM is, apart from being an NPM for browser based modules, actually just a meta-registry (allowing `package.json` to be augmented, and allowing short JSPM names to point to Github or NPM repositories that contain the actual code).
  8. Explain how `package.json` augmentation occurs.
  9. Explain the nature of the JSPM / SystemJS / ES6 Module Loader stack.
  10. Explain why SystemJS's `System.register()` makes it possible to have CJS, AMD, ES6 and Global modules all inter-dependent, including with circular dependencies, due to working in the same way that ES2016 browsers will.
  11. Explain why replacing browser-modules with SystemJS and MockableSystemJS will solve the intractable circular dependency problems we occasionally see.
  12. Explain why, if we were to have to have a hard dependency on Node.js, than we would use JSPM to do our bundling for us since it produces `System.register` format, which babelify also supports, but only creates anonymous modules.
  13. Explain why, if we were to have a soft dependency on Node.js, then we would require the `node_modules` directory to be pre-populated (either they downloaded some libraries in that format, or they use NPM themselves), and that we would need to scan each module's `package.json` for `dependencies`, which we then grabbed out of `node_modules`, and made available using `System.register()` and `System.config()`. 
  14. Explain why JSPM's flat module structure is so much more convenient than NPM's hierarchical module system (ability to break encapsulation if needed and means the smarts can be put into SystemJS, rather than requiring something complicated like browserify to produce the bundles), making having a soft dependency on Node.js at least conceivable.
  15. Explain how we might struggle to come up with _sound_ version names if producing the bundles ourselves, and how we may instead just have to pick arbitrary unique identifiers, making it harder for people to override when mocking.
  16. Explain how JSPM (like webpack) can also support non JavaScript assets, like CSS stylesheets and HTML templates, and that to do that for re-usable libraries you have to produce NPM modules that, at present, only a JSPM user could support, since there is presently no transpiler available to convert to ES5.
  17. Explain how JSPM only modules would need to take care not to use the `jspm` guard in `package.json`, since we'd be reliant on NPM to do our bundling for us.

**Comment by janhancic** [comment]
If you'll setup a hangout call or something, can I get in on that?

**Comment by dchambers** [comment]
Sounds good @janhancic, how about we get on a call at 3pm? It may take a couple of hours to go through all of this though, with a break in the middle, but I think your feedback would be really valuable, and we need to get this stuff right.

**Comment by dchambers** [comment]
Oh, @andyberry88 and @ioanalianabalas will also be dialling into that call.

**Comment by dchambers** [comment]
Some feedback from the meeting was that the concrete versions of the NPM libraries can be determined trivially (by reading `package.json`), and supporting non JavaScript assets is currently a half-baked idea (can no longer remember why unfortunately).

We also talked about the soft-dependency Vs hard-dependency trade-off, and I remembered that the majority of the work would be to create a Java implementation of `require.resolve()`, which seems like quite a small amount of work based on another [non Node.js implementation](https://www.npmjs.com/package/require-resolve) I've looked at.

Here is a rough break down of tasks:

**Switch to SystemJS:**
  1. Convert app from browser-modules to SystemJS.
  1. Make sub-realms work via MockableSystemJS.

**Add NPM Support:**
  1. Create Java version of `require.resolve()`.
  1. Create NPM bundler that behaves like this:
    1. Read seed dependencies for project's `package.json`.
    1. Use `require.resolve()` to find the concrete dependencies.
    1. Read `package.json` for each concrete dependency to determine the concrete version, and to determine its sub-dependencies.
    1. Again, use `require.resolve()` to find the concrete dependencies for the sub-dependencies, etc.

**Core ES6 Support:**

Regarding core ES6 Support, we shouldn't do anything until we are ready to accept a hard-dependency on Node.js. If/when this happens, there are two separate stories we might want to consider:
  1. Adding JSPM support so that we use ES6 binding semantics rather than binding to the transpiled ES5, and so that we can use modules that depend on non JavaScript assets.
  1. Adding ES6 support for non NPM classes.

**New Registry Support**:

The new registry would be able to proceed using SystemJS and MockableSystemJS.

**Comment by dchambers** [comment]
I've just been speaking to @andyberry88, who's thought of a few more issues with this proposal:

  1. NPM users will need to set a `NODE_PATH` environment variable pointing to the `node_modules` directory contained within the SDK.
  1. They will need their `package.json` to have entries for each of the SDK libraries currently updated to use NPM.
  1. Every time we update a library to NPM, this will cause a backwards compatibility issue.
  1. If they set the version of internal SDK libraries to anything other than `latest`, then this may cause them not to get updates when they update the SDK.

We decided that these problems could be solved by having a `brjs npm` command that did the following:

  1. Updated `package.json` to include valid entries for all of the SDK NPM libraries that the app uses.
  1. Set the `NODE_PATH` environment variable.
  1. Passed the received arguments to the underlying `npm` command, and let it do the right thing.

**Comment by dchambers** [comment]
I've been why I previously came to the conclusion that JSPM browser modules (containing non JavaScript assets) were half-baked was due to fact that the CSS plug-in is normally brought in from GitHub, and not NPM. However, I've since learned a couple of things:

  1. There is a version of the [JSPM CSS Plug-in](https://www.npmjs.com/package/system-css) available on NPM.
  2. Linking to that from NPM dependencies doesn't prevent the modules being used with JSPM since it automatically figures out the module is dependent on the CSS plug-in anyway, yet having this module available in `node_modules` should theoretically make it possible to use without needing JSPM to be there.

So, more investigation needed to prove this completely, but it looks promising so far...

**Comment by dchambers** [comment]
I've created a `systemjs-spike` branch of BRJS which almost seems like it will work, except I've ran into a problem due to the fact that unbundled code must now require its dependencies asynchronously, for example:

  1. Code within script tags (within aspect and workbench pages) isn't part of the bundle, but we expect both `require()` and global references to work.
  2. Tests aren't currently bundled either, so again `require()` and global references won't work.
  3. Even if tests were bundled, our plan to create a [MockableSystemJS](https://github.com/dchambers/MockableSystemJS) relies on us being able to re-register modules dynamically, yet this would again mean that we would subsequently need to asynchronously import modules, yet our current tests are all synchronous.

These problems could be solved if there was a synchronous way to load all modules, for example:

```js
System.instantiateRegisteredModules();
```

or alternatively:

```js
// this would throw if the module had not been pre-registered
System.instantiate('the-module');
```

I'll look into the feasibility of these options...

**Comment by dchambers** [comment]
I overcame the async problems by using [Bluebird](https://github.com/petkaantonov/bluebird) to allow me to create a `System.syncImport()` method. However, it's worth noting that if we use SystemJS, the ES6 Module Loader Polyfill & Bluebird together than we go from this:

  * Browser-Modules: 2KB (minified & gzipped)

to this:

  * SystemJS + ES6 Module Loader Polyfill: 16KB (minified & gzipped)
  * Minimal Bluebird Build: 11KB (minified & gzipped)

plus we force everybody to use Bluebird for their promises.

**Comment by dchambers** [comment]
I've run into a number of edge cases since switching over to SystemJS:

  1. The fell library didn't work because it's dependency on emitr was being processed by the wrong `require()` function, seemingly due to the fact that it seemed to have parts of browser-modules pre-sewn into it &mdash; this clearly needs fixing anyway.
  2. We previously had a stealth module called `app-meta!$data` written by a single plug-in, but that BRJS was not aware of, but with SystemJS it's no longer possible to hide dependencies from BRJS on the server but then still load them dynamically on the client &mdash; failing fast on hacks like this is a good thing, and should be fixed properly.
  3. I had to add some JavaScript so that BRJS/RequireJS style plug-ins (e.g. `plugin!resource-name`) could be supported without interfering with SystemJS style plugins (e.g. `resource-name!plugin`).

**Comment by dchambers** [comment]
Here is some code I used to client-side convert RequireJS style plug-ins (e.g. `plugin!require-path`) into something (e.g. `plugin#require-path`) that doesn't interfere with SystemJS style plug-ins (e.g. `require-path!plugin`):

```js
var origRegister = System.register;
System.register = function(name, deps, executingRequire, execute) {
	if(typeof(name) == 'string') {
		name = name.replace(/^(alias|service)!/, '$1#');
		deps = deps.map(function(depName) {
			return depName.replace(/^(alias|service)!/, '$1#');
		});
	}

	return origRegister.call(this, name, deps, executingRequire, execute);
};

var origNormalize = System.normalize;
System.normalize = function(name, parentName, parentAddress) {
	name = name.replace(/^(alias|service)!/, '$1#');
	return origNormalize.call(this, name, parentName, parentAddress);
};
```

Unfortunately this didn't work since the `require` that gets passed into CommonJS modules doesn't run through the System hooks, like _normalize_.

**Comment by dchambers** [comment]
By upgrading SystemJS from 0.16.x to 0.17.x the download size for SystemJS + ES6 Module Loader has gone from 16KB to 9.5KB.

**Comment by dchambers** [comment]
@andyberry88, I don't know if you need to pay particularly close attention, but I've updated the [MockableSystemJS](https://github.com/dchambers/MockableSystemJS) docs so it offers an equivalent API to what SystemJS itself offers, rather than trying to offer a 'prettier' API that might obscure some functionality. My thinking here is that other libraries could easily add a neater API on top, and in fact we will be required to continue offering a version of [sub-realms](https://github.com/BladeRunnerJS/browser-modules#sub-realms) that builds on top of MockableSystemJS.

**Comment by dchambers** [comment]
I'm closing this issue even though I haven't yet created a MockableSystem library, as there is now a separate issue for this.
